crates/api/src/linker.rs

```diff
@@ -0,0 +1,357 @@
+        let key = self.import_key(module, name, ty);
+        match self.map.entry(key) {
+            Entry::Occupied(o) => bail!(
+                "import of `{}::{}` with as {:?} defined twice",
+                module,
+                name,
+                o.key().kind,
+            ),
+            Entry::Vacant(v) => {
+                v.insert(item);
+            }
+        }
+        Ok(())
+    }
+
+    fn import_key(&mut self, module: &str, name: &str, ty: ExternType) -> ImportKey {
+        ImportKey {
+            module: self.intern_str(module),
+            name: self.intern_str(name),
+            kind: match ty {
+                ExternType::Func(f) => ImportKind::Func(f),
+                ExternType::Global(f) => ImportKind::Global(f),
+                ExternType::Memory(_) => ImportKind::Memory,
+                ExternType::Table(_) => ImportKind::Table,
+            },
+        }
+    }
+
+    fn intern_str(&mut self, string: &str) -> usize {
+        if let Some(idx) = self.string2idx.get(string) {
+            return *idx;
+        }
+        let string: Rc<str> = string.into();
+        let idx = self.strings.len();
+        self.strings.push(string.clone());
+        self.string2idx.insert(string, idx);
+        idx
+    }
+
+    /// Attempts to instantiate the `module` provided.
+    ///
+    /// This method will attempt to assemble a list of imports that correspond
+    /// to the imports required by the [`Module`] provided. This list
+    /// of imports is then passed to [`Instance::new`] to continue the
+    /// instantiation process.
+    ///
+    /// Each import of `module` will be looked up in this [`Linker`] and must
+    /// have previously been defined. If it was previously defined with an
+    /// incorrect signature or if it was not prevoiusly defined then an error
+    /// will be returned because the import can not be satisfied.
+    ///
+    /// # Errors
+    ///
+    /// This method can fail because an import may not be found, or because
+    /// instantiation itself may fail. For information on instantiation
+    /// failures see [`Instance::new`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn main() -> anyhow::Result<()> {
+    /// # let store = Store::default();
+    /// let mut linker = Linker::new(&store);
+    /// linker.func("host", "double", |x: i32| x * 2)?;
+    ///
+    /// let wat = r#"
+    ///     (module
+    ///         (import "host" "double" (func (param i32) (result i32)))
+    ///     )
+    /// "#;
+    /// let module = Module::new(&store, wat)?;
+    /// linker.instantiate(&module)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn instantiate(&mut self, module: &Module) -> Result<Instance> {
+        let mut imports = Vec::new();
+        for import in module.imports() {
+            let key = self.import_key(import.module(), import.name(), import.ty().clone());
```

Comment:
It's a little odd to me that `Linker::instantiate` needs to be `&mut self`.  Perhaps an `Linker::insert_key` and `Linker::get_key` method, where the former does what `Linker::import_key` is doing now and the latter returns `Option<ImportKey>` after checking if the given strings are present in the data structures without interning?  That would remove the requirement of this being `&mut self`.

Reply:
In general builders in Rust have `&mut self` even on the build methods, primarily for future-flexibility. That being said given the way I envision this being used I think we'll want a guarantee that `&mut self` isn't needed and that mutations don't happen. In that sense I agree, let's shoot for `&self` here.